src/meta-srv/src/selector/load_based.rs

```diff
@@ -12,22 +12,120 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::HashMap;
+
 use api::v1::meta::Peer;
 use common_meta::key::TableMetadataManager;
 use common_meta::rpc::router::find_leaders;
-use common_telemetry::warn;
+use common_telemetry::debug;
+use parking_lot::RwLock;
 use snafu::ResultExt;
 use table::metadata::TableId;
 
 use crate::error::{self, Result};
-use crate::keys::{LeaseKey, LeaseValue, StatKey};
+use crate::keys::{LeaseKey, LeaseValue, StatKey, StatValue};
 use crate::lease;
 use crate::metasrv::SelectorContext;
-use crate::selector::{Namespace, Selector};
+use crate::selector::common::choose_peers;
+use crate::selector::weight_compute::{RegionNumsBasedWeightCompute, WeightCompute};
+use crate::selector::weighted_choose::{RandomWeightedChoose, WeightedChoose};
+use crate::selector::{Namespace, Selector, SelectorOptions};
+
+pub struct LoadBasedSelector<W, C> {
+    weighted_choose: RwLock<W>,
+    weight_compute: C,
+}
+
+impl<W, C> LoadBasedSelector<W, C> {
+    pub fn new(weighted_choose: W, weight_compute: C) -> Self {
+        Self {
+            weighted_choose: RwLock::new(weighted_choose),
+            weight_compute,
+        }
+    }
+}
 
-const MAX_REGION_NUMBER: u64 = u64::MAX;
+impl Default for LoadBasedSelector<RandomWeightedChoose<Peer>, RegionNumsBasedWeightCompute> {
+    fn default() -> Self {
+        Self {
+            weighted_choose: RwLock::new(RandomWeightedChoose::default()),
+            weight_compute: RegionNumsBasedWeightCompute,
+        }
+    }
+}
 
-pub struct LoadBasedSelector;
+#[async_trait::async_trait]
+impl<W, C> Selector for LoadBasedSelector<W, C>
+where
+    W: WeightedChoose<Peer>,
+    C: WeightCompute<Source = HashMap<StatKey, StatValue>>,
+{
+    type Context = SelectorContext;
+    type Output = Vec<Peer>;
+
+    async fn select(
+        &self,
+        ns: Namespace,
+        ctx: &Self::Context,
+        opts: SelectorOptions,
+    ) -> Result<Self::Output> {
+        // 1. get alive datanodes.
+        let lease_kvs =
+            lease::alive_datanodes(ns, &ctx.meta_peer_client, ctx.datanode_lease_secs).await?;
+
+        // 2. get stat kvs and filter out expired datanodes.
+        let stat_keys = lease_kvs.keys().map(|k| k.into()).collect();
+        let stat_kvs = filter_out_expired_datanode(
+            ctx.meta_peer_client.get_dn_stat_kvs(stat_keys).await?,
+            &lease_kvs,
+        );
+
+        // 3. make sure the regions of a table are distributed on different datanodes.
+        let stat_kvs = if let Some(table_id) = ctx.table_id {
+            let table_metadata_manager = TableMetadataManager::new(ctx.kv_backend.clone());
+
+            let leader_peer_ids = get_leader_peer_ids(&table_metadata_manager, table_id).await?;
+
+            filter_out_datanode_by_table(stat_kvs, &leader_peer_ids)
```

Comment:
Imagine the following scenario: there are 2 datanodes, a table has 3 regions, and different regions of this table inevitably have to be distributed on the same datanode. At this time, `filter_out_datanode_by_table` returns an empty set, which is not what we want. My suggestion is to use the original `stat_kvs` as a return value when the result of `filter_out_datanode_by_table` is an empty set (which means: we will do our best to distribute regions evenly, but if the conditions are not met, we will assign multiple regions to the same datanode).

Reply:
In the scenario of region failover, we also need to avoid selecting failed datanode (that is, the datanode where the failed region id is located).